### PR TITLE
AK/MIN-259 Rework risk-manager offchain worker.

### PR DESCRIPTION
**Pull request by kurotych**
**Description:**

Removed randomness in Liquidation check
Added max duration time for check loans procedure. 
Added storage lock to liquidation-pools worker to prevent any race condition.

If deadline is reached then next check loans procedure will start from last processed pool.
For example:
Initial queue [BTC, ETH, DOT, KSM]
- Worker starts processing loans for block N. Deadline was reached when worker was processing DOT pool 
- Worker starts processing loans for block N+1. Queue will looks like: [DOT, KSM, BTC, ETH]

_If the deadline is not reached, then all pools have been processed._ 

P.S. I think MaxDurationTime for worker should be less than block production time. It will prevent asynchronous problems.
P.S.S. OffchainWorkerMaxDurationMs is applicable only for loans check procedure. 
I hope pool balancing procedure won't take a lot of time. 

**Changes in storage or API:**

Removed storage value:
`BalancingPeriod<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;` 
Removed Extrinsic: 
`setBalancingPeriod`

**Checklist**

- [x] Code is covered with tests
- [x] Code is covered with benchmarks
- [x] Fresh benchmark results included

### CODE REVIEW

File: pallets/liquidation-pools/src/lib.rs [hunk 1]
```diff
@@ -53,6 +54,7 @@ pub struct LiquidationPoolData {
 }
 
 type BalanceResult = sp_std::result::Result<Balance, DispatchError>;
+type DispatchResult = sp_std::result::Result<(), DispatchError>;
```
Comment (lut1k) [inline pallets/liquidation-pools/src/lib.rs:57]: WYF?! :D

File: pallets/liquidation-pools/src/lib.rs [hunk 2]
```diff
@@ -76,10 +78,6 @@ pub mod module {
 		/// The Liquidation Pool's module id, keep all assets in Pools.
 		type LiquidationPoolsModuleId: Get<ModuleId>;
 
-		#[pallet::constant]
-		/// The Liquidation Pool's account id, keep all assets in Pools.
-		type LiquidationPoolAccountId: Get<Self::AccountId>;
```
Comment (lut1k) [inline pallets/liquidation-pools/src/lib.rs:81]: This constant is currently used by our frontend. If we delete it, the frontend will fall off. We can expose the ModuleId, but then the frontend will have to transform the ModuleId into an AccountId.

File: pallets/traits/src/lib.rs [hunk 3]
```diff
@@ -37,6 +37,10 @@ pub trait PoolsManager<AccountId> {
 	fn pool_exists(underlying_asset: &CurrencyId) -> bool;
 }
 
+pub trait LiquidationPoolsManager<AccountId>: PoolsManager<AccountId> {
+	fn pools_balancing() -> DispatchResult;
```
Comment (lut1k) [inline pallets/traits/src/lib.rs:41]: Add explanatory comments.

File: pallets/risk-manager/src/mock.rs [hunk 4]
```diff
@@ -61,11 +64,40 @@ parameter_types! {
...
+	type LockOrigin = EnsureSignedBy<ZeroAdmin, AccountId>;
+	type WeightInfo = ();
+}
+
 pub struct WhitelistMembers;
 mock_impl_system_config!(Test);
 mock_impl_orml_tokens_config!(Test);
 mock_impl_orml_currencies_config!(Test);
-mock_impl_liquidity_pools_config!(Test);
+mock_impl_liquidity_pools_config!(Test, Prices);
```
Comment (lut1k) [inline pallets/risk-manager/src/mock.rs:100]: For the liquidity-pools pallet we use the prices pallet, and for the liquidation-pools pallet we use `struct MockPriceSource`. 
It should be?

File: pallets/risk-manager/src/lib.rs [hunk 5]
```diff
@@ -113,6 +107,10 @@ pub mod module {
 		type RiskManagerUpdateOrigin: EnsureOrigin<Self::Origin>;
 
 		type RiskManagerWeightInfo: WeightInfo;
+
+		#[pallet::constant]
```
Comment (lut1k) [inline pallets/risk-manager/src/lib.rs:111]: Do we really need to expose this parameter?
https://substrate.dev/docs/en/knowledgebase/runtime/macros#palletconstant

File: pallets/liquidation-pools/src/lib.rs (hunk 1 above)
Comment (kurotych) [reply pallets/liquidation-pools/src/lib.rs]: Code duplication. nothing special. Removed)

File: pallets/liquidation-pools/src/lib.rs (hunk 2 above)
Comment (kurotych) [reply pallets/liquidation-pools/src/lib.rs]: Restored

File: pallets/traits/src/lib.rs (hunk 3 above)
Comment (kurotych) [reply pallets/traits/src/lib.rs]: Done

File: pallets/risk-manager/src/lib.rs (hunk 5 above)
Comment (kurotych) [reply pallets/risk-manager/src/lib.rs]: Removed

File: pallets/risk-manager/src/mock.rs (hunk 4 above)
Comment (kurotych) [reply pallets/risk-manager/src/mock.rs:100]: Yes, we test risk-manager related code here. There is no matter what pallet is used in liquidation

File: pallets/risk-manager/src/lib.rs [hunk 6]
```diff
@@ -358,99 +356,110 @@ pub mod module {
...
 
-		let to_be_continue = StorageValueRef::persistent(&OFFCHAIN_WORKER_DATA);
-
-		// Get to_be_continue record
-		let (collateral_position, start_key) =
-			if let Some(Some((last_collateral_position, maybe_last_iterator_previous_key))) =
-				to_be_continue.get::<(u32, Option<Vec<u8>>)>()
-			{
-				(last_collateral_position, maybe_last_iterator_previous_key)
+		let latest_pool_index = {
```
Comment (lut1k) [inline pallets/risk-manager/src/lib.rs:372]: `.ok_or(OffchainErr::CheckFail)? as usize;` - this line will return `2`. Why?

Suggestion:
```
let start_pool_index = match StorageValueRef::persistent(&OFFCHAIN_WORKER_LATEST_POOL_INDEX)
.get::<u32>() {
	Some(Some(index)) => {
		// Assume that count of enbled tokens can be changed. So make sure that index is not out of
		// bounds
		(index as usize % underlying_assets.len())
	}
	_ => usize::zero(),
};
```
I left checking the index. But it is not clear how the index can be greater than the length of the vector. Only if the currency is removed?

File: pallets/risk-manager/src/lib.rs [hunk 7]
```diff
@@ -358,99 +356,110 @@ pub mod module {
...
-
-			iteration_count += 1;
-
-			// extend offchain worker lock
-			guard.extend_lock().map_err(|_| OffchainErr::OffchainLock)?;
+			debug::info!("RiskManager finished processing loans for {:?}", currency_id);
 		}
 
-		let iteration_end_time = sp_io::offchain::timestamp();
+		let working_time = sp_io::offchain::timestamp().diff(&working_start_time);
```
Comment (lut1k) [inline pallets/risk-manager/src/lib.rs:444]: Above this line, I would add:
`StorageValueRef::persistent(&OFFCHAIN_WORKER_LATEST_POOL_INDEX).clear();`

File: pallets/risk-manager/src/lib.rs [hunk 8]
```diff
@@ -358,99 +356,110 @@ pub mod module {
...
+						debug::info!(
+							"Risk Manager offchain worker hasn't(!) processed all pools. \
+							 MAX duration time is expired. Loans checked count: {:?}, \
+							 loans liquidated count: {:?}",
+							loans_checked_count,
+							loans_liquidated_count
+						);
+						StorageValueRef::persistent(&OFFCHAIN_WORKER_LATEST_POOL_INDEX).set(&(pos as u32));
+						return Ok(());
+					}
```
Comment (lut1k) [inline pallets/risk-manager/src/lib.rs:438]:
```suggestion
guard.extend_lock().map_err(|_| {
	// The lock's deadline is happened
	debug::info!(
		"Risk Manager offchain worker hasn't(!) processed all pools. \
			 MAX duration time is expired. Loans checked count: {:?}, \
			 loans liquidated count: {:?}",
		loans_checked_count,
		loans_liquidated_count
	);
	StorageValueRef::persistent(&OFFCHAIN_WORKER_LATEST_POOL_INDEX).set(&(pos as u32));
	OffchainErr::OffchainLock
})?;
```

File: pallets/risk-manager/src/tests.rs [hunk 9]
```diff
@@ -6,6 +6,127 @@ use mock::{Event, *};
...
+		let ex: Extrinsic = Decode::decode(&mut &*transaction).unwrap();
+
+		// Called extrinsic input params
+		let (who, pool_id) = match ex.call {
+			crate::mock::Call::TestRiskManager(crate::Call::liquidate(who, pool_id, ..)) => (who, pool_id),
+			e => panic!("Unexpected call: {:?}", e),
+		};
+		assert_eq!(who, ALICE);
+		assert_eq!(pool_id, KSM);
+	});
```
Comment (lut1k) [inline pallets/risk-manager/src/tests.rs:127]: I think it is necessary to add a check of the local storage that it is empty.

File: pallets/risk-manager/src/lib.rs (hunk 6 above)
Comment (kurotych) [reply pallets/risk-manager/src/lib.rs:372]: - Only if the currency is removed?
- Yes it is.

File: pallets/risk-manager/src/lib.rs (hunk 7 above)
Comment (kurotych) [reply pallets/risk-manager/src/lib.rs:444]: I prefer to clear this storage right away after get value from it.

File: pallets/risk-manager/src/tests.rs (hunk 9 above)
Comment (kurotych) [reply pallets/risk-manager/src/tests.rs:133]: Done

File: pallets/risk-manager/src/lib.rs (hunk 8 above)
Comment (kurotych) [reply pallets/risk-manager/src/lib.rs]: Done

**Review by lut1k — APPROVED** [review]

**Review by lut1k — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**